pkg/util/iptree/iptree.go

```diff
@@ -0,0 +1,602 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptree
+
+import (
+	"fmt"
+	"math/bits"
+	"net/netip"
+	"reflect"
+)
+
+// based on https://github.com/armon/go-radix but
+// the tree is binary to simplify implementation.
+
+// node is an element of radix tree with a netip.Prefix optimized to store IP Addresses.
+type node[T any] struct {
+	// prefix network CIDR
+	prefix netip.Prefix
+
+	// leaf is used to store value
+	leaf bool
+	val  T
+
+	child [2]*node[T]
+}
+
+func (n *node[T]) isLeaf() bool {
+	return n.leaf
+}
+
+func (n *node[T]) mergeChild() {
+	// can not merge if there are two or no childs
```

Comment:
is a leaf if there are no children, right?

Reply:
But it can also be a leaf even if there are children, can't it? (Which is a weird definition of "leaf", but whatevs.)

e.g., the check at line 461 assumes that `parent` can both have a child and be a leaf (and that it shouldn't be merged if it's a leaf).

Reply:
leaf is confusing indeed, it really means "node that can be a leaf because it contains a value" or "node that is not added by the user"

Reply:
I will switch it to public , as it is a node that is set by the user